controller/engine_controller.go

```diff
@@ -1814,6 +1828,63 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replica, addr st
 	return nil
 }
 
+// updateReplicaRebuildFailedCondition updates the rebuild failed condition if replica rebuilding failed
+func updateReplicaRebuildFailedCondition(log *logrus.Entry, ds *datastore.DataStore, replicaName, errMsg string) error {
+	replica, err := ds.GetReplica(replicaName)
+	if err != nil {
+		return err
+	}
+
+	replicaRebuildFailedReason, conditionStatus, err := getReplicaRebuildFailedReason(ds, replica.Spec.NodeID, errMsg)
+	if err != nil {
+		return err
+	}
+
+	replica.Status.Conditions = types.SetCondition(
+		replica.Status.Conditions, longhorn.ReplicaConditionTypeRebuildFailed,
+		conditionStatus, replicaRebuildFailedReason, errMsg)
+
+	_, err = ds.UpdateReplicaStatus(replica)
+
+	return err
+}
+
+func getReplicaRebuildFailedReason(ds *datastore.DataStore, replicaNodeID, errMsg string) (string, longhorn.ConditionStatus, error) {
+	replicaRebuildFailedReason, conditionStatus, isRebuildingFailedByNetwork := getReplicaRebuildFailedReasonFromError(errMsg)
+	if isRebuildingFailedByNetwork {
+		replicaKubernetesNode, err := ds.GetKubernetesNode(replicaNodeID)
+		if err != nil {
+			return "", "", err
+		}
+		for _, con := range replicaKubernetesNode.Status.Conditions {
+			if con.Type != v1.NodeReady {
```

Comment:
Suggest using longhorn node resource condition instead to avoid duplicated query. 

Also, NotReady doesn't mean `NodeDown`.

Reply:
Removed getting KubernetesNode, and use longhorn node resource condition instead. Thanks